### ISSUE TITLE
New Check: IllegalSymbol to forbit emoj in code

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#18065]
AI generates unicode emoj in code comments
[example](https://github.com/checkstyle/checkstyle/pull/18008/files#diff-e8685d13c843aea432b0c89f0ef7f09fddfd9f00bf2c332e7e7d1a47e776354dR60): 

> // ✅ Enhancement 1: Basic validation

this also raised by other users to address this at https://github.com/checkstyle/checkstyle/discussions/17983#discussioncomment-14888080

Potential soluton: 
```
    <module name="RegexpSingleline">
        <property name="format" value="\u2705" />
        <property name="message" value="Unicode emoj symbols should not be used." />
    </module>
```

use https://www.mauvecloud.net/charsets/CharCodeFinder.html in mode "Hexadecimal Character Codes" to get code.

To make it easier to search and collaborate to extend to all other Unicode pictures, we can put this as example https://checkstyle.org/checks/regexp/regexpsingleline.html#Examples so it will part of git files and user can contribute to make list to cover most

Website sources are part of main repository https://github.com/checkstyle/checkstyle/blob/master/src/site/xdoc/checks/regexp/regexpsingleline.xml.template

We need to put in https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle-checks.xml config  that violates majority of unicode emoj symbols, try to find in web or AI list of them.

**Comment by romani** [comment]
@zyadahmed, can you help us fix it ?

**Comment by zyadahmed** [comment]
@romani sure

**Comment by zyadahmed** [comment]
@romani  After multiple attempts ,it turns out that reliably detecting all emojis is extremely difficult. Java’s regex  operates on UTF-16 code units rather than full Unicode codepoints, so many emojis cannot be matched consistently. Because of these technical limitations, we cannot accurately catch most emojis using a simple or stable regex-based Checkstyle rule, so what is your recommendation ?

**Comment by romani** [comment]
Nice reason to create special Check that will forbid usage of special symbols by codes or ranges of codes, I hope in utf table such symbols are in special ranges of codes , so it will be easier to mark them as forbidden.

Please suggest name of Check , and what properties you think should be there, and draw by text example what could be a behavior of Check.
We usually try to agree upfront on design and behavior to avoid wasting contributors time to redo and rename in mid of development.

**Comment by zyadahmed** [comment]
We have many combinations of emojis and special characters. Some exist in fixed Unicode blocks, while others are scattered across the utf  table and keep expanding over time. Because of this, hard-coding all ranges isn’t practical. But we can still detect the most common ranges used by AI tools and developers. A simple Check may named `SpecialSymbolsCheck.java` can cover the most frequent emojior symbol blocks without tracking every new Unicode update .

**Comment by romani** [comment]
Majority of code doesn't have Unicode symbols at all, ASCII is kind of standard for writing of code. So we can exclude by big ranges. If some runs in conflict, he will update config with more precise ranges.

Checks should be able to catch character by code, no it will be general. Some will use it for emojis others for no fancy punctuations, not Japan period symbol, .....
All this can be shared in Examples of Check.

**Comment by zyadahmed** [comment]
So we can implement a check whose main logic is to verify if a character is part of ASCII, and the user can customize the allowed ranges based on their needs. We can also provide examples with other language.

**Comment by romani** [comment]
I see this Check as IllegalSymbol that validates special symbol by code should not happen in some token.
Imagination:
```
<module name="IllegalSymbol">
        <property name="symbolCodes" value="\u2705" />
<property name="token" value="COMMENT" />
        <property name="message" value="Unicode emoj symbols should not be used in comment" />
    </module>
```

```
<module name="IllegalSymbol">
        <property name="symbolCodes" value="0xd83c-0xd83e, 0xdd25, 0xd83d, 0xde0a, 0xdf81" />
<property name="token" value="COMMENT" />
        <property name="message" value="emoj symbols should not be used in comment" />
    </module>
```

**Comment by zyadahmed** [comment]
I think that would be a great design, and in the Shared Check example we can highlight the preferred way to use  ASCII characters only

**Comment by romani** [comment]
It can be default setup of Check , just ASCII codes allowed.

**Comment by aclfe** [comment]
@romani 

I noticed this discussion on banning emojis in code comments and it relates to my current progress.  

I've currently got a local working build for issue #10629. 
See comment for progress: https://github.com/checkstyle/checkstyle/issues/10629#issuecomment-3721923411

I just wanted to confirm from an architectural point of view, I assume the Javadoc Grammar should remain permissive and fully support parsing these Unicode characters while the actual restriction of emojis in code comments should be handled by 
IllegalSymbol? Futhermore, **would this check also extend to Javadocs**?

**Comment by romani** [comment]
Grammar stays unchanged.

Special Check does validation of such symbols presence, user always has rights to use this check when such symbols are not good.

New Check should look at javadoc comments and any other comments.

**Comment by aclfe** [comment]
Understandable. 

I also propose an easier implementation for this check. The discussion mentions that it's hard to work with emojis, but there is already good coverage in JDK21. 

Please see https://inside.java/2023/11/20/sip089/ 

It introduces methods like 
```java
public static boolean isEmoji(int codePoint)
public static boolean isEmojiPresentation(int codePoint)
public static boolean isEmojiModifier(int codePoint)
```
that help give better coverage of emojis. As we know, Checkstyle has already migrated to JDK21, we can utilise these methods instead.

> As the usage of emojis has become increasingly common, JDK-8303018 was created to add API to java.lang.Character to give Java developers the ability to check if emojis were being used

**Comment by smita1078** [comment]
@romani  Please review when you get some time

**Comment by romani** [comment]
Check is merged.
